### ISSUE TITLE
[AArch64] Incorrect copy propagation across BLR_RVMARKER

### ISSUE THREAD

**Issue by kyulee-com** [llvm/llvm-project#61907]
`BLR_RVMARKER` is a pseudo instruction attaching ARC related helpers + operations.
In the following cutdown, given two instances, `mov x0, x19` before and after `BLR_RVMARKER`, the latter is optimized by machine copy propagation (machine-cp), which caused mis-compilation along with other optimizations in our code base.

```
$ cat test.ll
target triple = "arm64-apple-ios13"

%0 = type opaque
%1 = type opaque
%2 = type opaque

define %0* @foo(i8* %0) {
entry:
  tail call void @f1(i8* null)
  %1 = bitcast i8* %0 to %0*
  %call1 = call %0* @f2(%0* %1) [ "clang.arc.attachedcall"(i8* (i8*)* @llvm.objc.retainAutoreleasedReturnValue) ]
  tail call void @f3(i8* %0)

  ret %0* null
}

declare void @f1(i8*)
declare void @f3(i8*)

define %0* @f2(%0* noundef returned %dict) {
entry:
  ret %0* null
}

declare i8* @llvm.objc.retainAutoreleasedReturnValue(i8*)
```

```
$ llc < test.ll
...
	mov	x0, x19
	bl	_f2
	mov	x29, x29
	bl	_objc_retainAutoreleasedReturnValue
	// mov	x0, x19  ; This copy is deleted by machine-cp
	bl	_f3
...
```

I figured `BLR_RVMARKER` sometimes does not set `implicit-def $x0` as it inherits from the first call, which allows this transformation valid.
Since `BLR_RVMARKER` combines a objc helper call like `objc_retainAutoreleasedReturnValue` which always have a return value, I wonder if we should conservatively add `X0` to `Defs` to this instruction in `.td` file.
Or, this is a valid optimization, which we should allow?

**Comment by fhahn** [comment]
This looks like a bug and would be good to fix. cc @ahatanak @ahmedbougacha

**Comment by ahmedbougacha** [comment]
I think what makes this okay is the `returned` attribute on `f2`.   So the whole `BLR_RVMARKER` ends up with `CSR_Darwin_AArch64_AAPCS_ThisReturn_RegMask` (which preserves `x0`) instead of the usual `CSR_Darwin_AArch64_AAPCS_RegMask`.  So wouldn't it be correct to eliminate that copy, since both calls in the `BLR_RVMARKER` return the same `x0` that was copied from `x19` before the calls?

I don't know if that has interesting downstream effects with other optimizations though.

But the inheriting of CCs/regmasks is indeed somewhat fishy, the call could have something weirder than just `ThisReturn`, in which case neither call's regmask would be correct (the first weird one, or the second standard one, to the helper function), and we'd probably need to combine them somehow?
I'm not sure we can do that in the general case, there's a somewhat implicit contract in our usage of `BLR_RVMARKER` that `x0` is passed between the two calls, of course.

**Comment by kyulee-com** [comment]
I agree this specific case might be (luckily?) okay with the `returned` attribute on this copy propagation, but it was really bad on our downstream pass that moved around the copy `mov x0, x19` across this call boundary as `BLR_RVMARKER` does not interfere `x0` which should inherit from `objc_retainAutoreleasedReturnValue`.
I think we should conservatively add all implicit or bitmask operands of two actual calls into this pseudo `BLR_RVMARKER`.
